src/octopath.cpp

```diff
@@ -106,199 +153,333 @@ bool OctoClass::OctoRRG(const Eigen::Vector3d &p0,
+  *plan_time = (ros::Time::now() - t0).toSec();
 
-        // Check if new node connects with the final destination
-        // This portion does not need to execute after one path has been found between p0 and pf
-        cost = obj_rrg.DistanceToNode(index, pf);
-        if (cost <= max_dist) {
-            if (free_space_only) {
-                col_check = CheckCollision(pf, sample);
-            } else {
-                col_check = CheckOccupancy(pf, sample);
-            }
-
-            if (col_check != 1) {
-                obj_rrg.AddNode(pf, &final_index);
-                obj_rrg.AddEdge(index, final_index, cost);
-                connected_graph = true;
-                ROS_INFO("[mapper] Found connection to destination!");
-            }
-        }
-    }
+  // ROS_INFO("[mapper] nNodes: %d", static_cast<int>(obj_rrg.rrgraph_.n_nodes_));
+  // ROS_INFO("[mapper] nEdges: %d", static_cast<int>(obj_rrg.rrgraph_.n_edges_));
+
+  return true;
+}
 
-    // Publish graph into Rviz if requested
-    if (publish_rviz) {
-        obj_rrg.rrgraph_.GraphVisualization(graph_markers);
+bool OctoClass::Astar(const octomap::point3d &p0,
+                      const octomap::point3d &pf,
+                      const bool &prune_result,
+                      double *plan_time,
+                      std::vector<Eigen::Vector3d> *path,
+                      std::vector<Eigen::Vector3d> *pruned_path) {
+  const ros::Time t0 = ros::Time::now();
+
+  // Check whether p0 and pf are free nodes in the octomap
+  int is_occ;
+  is_occ = this->CheckOccupancy(p0);
+  if (is_occ == -1) {
+    ROS_INFO("[mapper] A* failed: initial node is unknown in the octomap!");
+    return false;
+  } else if (is_occ == 1) {
+    ROS_INFO("[mapper] A* failed: initial node is occupied in the octomap!");
+    return false;
+  }
+  is_occ = this->CheckOccupancy(pf);
+  if (is_occ == -1) {
+    ROS_INFO("[mapper] A* failed: final node is unknown in the octomap!");
+    return false;
+  } else if (is_occ == 1) {
+    ROS_INFO("[mapper] A* failed: final node is occupied in the octomap!");
+    return false;
+  }
+
+  // Get map info
+  IndexedKeySet indexed_free_keys;
+  std::vector<double> node_sizes;
+  Eigen::Vector3d map_min, map_max;
+  tree_inflated_.getMetricMin(map_min[0], map_min[1], map_min[2]);
+  tree_inflated_.getMetricMax(map_max[0], map_max[1], map_max[2]);
+  this->BBXFreeNodes(map_min, map_max, &indexed_free_keys, &node_sizes);
+  ROS_INFO("[mapper] A* computing trajectory from [%.3f, %.3f, %.3f] to [%.3f, %.3f, %.3f]",
+           p0.x(), p0.y(), p0.z(), pf.x(), pf.y(), pf.z());
+  ROS_DEBUG("[mapper] Number of nodes: %d", static_cast<int>(indexed_free_keys.Size()));
+
+  // Find initial and final keys/indexes
+  const octomap::OcTreeKey initial_key = tree_inflated_.coordToKey(p0);
+  const octomap::OcTreeKey final_key = tree_inflated_.coordToKey(pf);
+
+  uint initial_index;
+  if (!indexed_free_keys.Key2Index(initial_key, &initial_index)) {
```

Comment:
I think so... we are frequently checking distances to obstacles and free neighbors, we better plan the path in a static map.
If you are concerned that there is no mutex, the mutex can be found in `services.cpp`, as we call `globals_.octomap.Astar(...)`

Reply:
OK. If we wanted to make sure it was always mutexed, I would put the mutex in this function, otherwise it may be possible to call the function without mutexing, which we really don't want

Reply:
So here's the overall structure to `mapper`:
- There's the `OctoClass`, which contains the maps, relevant map functions, path planning, map updating, etc. `OctoClass` is a standalone class, and it has no threads.
- There's the `MapperClass`, which has all publishers/subscribers/services and threads. `MapperClass` has `OctoClass` as a member variable, and every time we do anything with the `OctoClass` we throw in a mutex to protect it. `OctoClass` interacts with the outside world (ROS) through `MapperClass`, and that's why I keep all mutexes in `MapperClass`.

Does it make sense?

This whole code was written 3 years ago, and it was really my first experience with a large c++ development. I know that it isn't ideal, and I would definitely do things differently today...

Reply:
I see, if all the updates to the map come through MapperClass then it makes sense for the mutexes to live there, since that's where the asynchronous behavior comes from